reflect.go

```diff
@@ -0,0 +1,387 @@
+
+	// find all unchecked APIs to add them to checkedAPIs after the pass
+	notCheckedAPIs := make(map[string]bool)
+	for _, knownAPI := range maps.Keys(cachedOutput.KnownReflectAPIs) {
+		if !checkedAPIs[knownAPI] {
+			notCheckedAPIs[knownAPI] = true
+		}
+	}
+
+	tf.ignoreReflectedTypes(ssaPkg)
+
+	// all previously unchecked APIs have now been checked add them to checkedAPIs,
+	// to avoid checking them twice
+	maps.Copy(checkedAPIs, notCheckedAPIs)
+
+	// if a new reflectAPI is found we need to Re-evaluate all functions which might be using that API
+	if len(cachedOutput.KnownReflectAPIs) > lenPrevKnownReflectAPIs {
+		tf.recordReflection(ssaPkg)
+	}
+}
+
+func (tf *transformer) ignoreReflectedTypes(ssaPkg *ssa.Package) {
+	for _, memb := range ssaPkg.Members {
+		if t, ok := memb.(*ssa.Type); ok {
+			// methods aren't package members only their reciever types are
+			// so some logic is required to find the methods a type has
+
+			// yes, finding all methods really only works with both loops
+			mset := ssaPkg.Prog.MethodSets.MethodSet(t.Type())
+			for i, n := 0, mset.Len(); i < n; i++ {
+				fun := ssaPkg.Prog.MethodValue(mset.At(i))
+				if fun != nil {
+					tf.checkCalls(fun)
+				}
+			}
+
+			mset = ssaPkg.Prog.MethodSets.MethodSet(types.NewPointer(t.Type()))
+			for i, n := 0, mset.Len(); i < n; i++ {
+				fun := ssaPkg.Prog.MethodValue(mset.At(i))
+				if fun != nil {
+					tf.checkCalls(fun)
+				}
+			}
+		}
+
+		if fun, ok := memb.(*ssa.Function); ok {
+			// these not only include top level functions, but also synthetic
+			// functions like the initialization of global variables
+			tf.checkCalls(fun)
+		}
+	}
+}
+
+func (tf *transformer) checkCalls(fun *ssa.Function) {
+	if fun.Synthetic == "loaded from gc object file" {
+		// fun.WriteTo crashes otherwise
+		return
+	}
+
+	/* fun.WriteTo(os.Stdout) */
+
+	var reflectParams []int
+
+	for _, block := range fun.Blocks {
+		for _, inst := range block.Instrs {
+			/* 	fmt.Printf("inst: %v, t: %T\n", inst, inst) */
+			call, ok := inst.(*ssa.Call)
+			if !ok {
+				continue
+			}
+			/* fmt.Printf("call: %v\n", call) */
+
+			callName := call.Call.Value.String()
+			if checkedAPIs[callName] {
+				// only check apis which were not already checked
+				continue
+			}
+
+			// record each call argument passed to a function parameter which is used in reflection
+			knownParams := cachedOutput.KnownReflectAPIs[callName]
+			for _, knownParam := range knownParams {
+				if len(call.Call.Args) <= knownParam {
+					continue
+				}
+
+				arg := call.Call.Args[knownParam]
+
+				/* fmt.Printf("flagging arg: %v\n", arg) */
+
+				visited := make(map[ssa.Value]bool)
+				reflectedParam := tf.recordArgReflected(arg, visited)
+				if reflectedParam == nil {
+					continue
+				}
+
+				pos := slices.Index(fun.Params, reflectedParam)
+				if pos < 0 {
+					continue
+				}
+
+				/* fmt.Printf("recorded param: %v func: %v\n", pos, fun) */
+
+				if !slices.Contains(reflectParams, pos) {
```

Comment:
I guess this linear search is OK because, typically, functions won't have more than a dozen or so parameters.